src/lib/Util/adapters/bbc-kaldi/group-words-by-speakers.js

```diff
@@ -0,0 +1,118 @@
+/**
+edge cases
+- more segments then words - not an issue if you start by matching words with segment
+and handle edge case where it doesn't find a match
+- more words then segments - orphan words
+ */
+function groupWordsInParagraphsBySpeakers(words, segments) {
+  // add speakers to each word
+  const wordsWithSpeakers = addSpeakerToEachWord(words, segments.segments);
+  // group words by speakers sequentially
+  const result = groupWordsBySpeaker(wordsWithSpeakers);
+
+  return result;
+};
+
+/**
+* Add speakers to each words
+* if it doesn't have add unknown attribute `U_UKN`
+* @param {*} words
+* @param {*} segments
+*/
+function addSpeakerToEachWord(words, segments) {
+  const tmpWordsWithSpeakers = [];
+  words.forEach((word) => {
+    const tmpSpeakerSegment = findSegmentForWord(word, segments);
+
+    word.speaker = formatSpeakerName(tmpSpeakerSegment.speaker);
+    tmpWordsWithSpeakers.push(word);
+  });
+
+  return tmpWordsWithSpeakers;
+}
+
+/**
+ * Groups Words by speaker attribute
+ * @param {array} wordsWithSpeakers - same as kaldi words list but with a `speaker` label attribute on each word
+ * @return {array} - list of paragraph objcts, with words, text and sepaker attributes.
+ * where words is an array and the other two are strings.
+ */
+function groupWordsBySpeaker(wordsWithSpeakers) {
```

Comment:
since this is returning paragraph level objects, is it not better to change the name of the function to be more descriptive?

It can be confusing because you pass around a bunch of arrays of words called similar things.

Reply:
leaving it for now, needs a more careful thought thought through naming convention coz there is `groupWordsInParagraphsBySpeakers` that calls helper function `groupWordsBySpeaker` etc..

and there's the destination as the draftJs paragraphs, and the intermediate ( throw away ) representations used to get from the kaldi STT words to that